Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp

```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
     });
 }
 
-static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILiteral uti, std::span<const unsigned> lengths)
+static Vector<RefPtr<ShareableBitmap>> bitmapsFromNativeImage(NativeImage& image, std::span<const unsigned> lengths)
 {
-    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceSRGB));
-    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
-    RetainPtr cfUTI = uti.createCFString();
-    RetainPtr destination = adoptCF(CGImageDestinationCreateWithData(destinationData.get(), cfUTI.get(), lengths.size(), nullptr));
+    Vector<RefPtr<ShareableBitmap>> bitmaps;
+    auto sourceColorSpace = image.colorSpace();
+    auto destinationColorSpace = sourceColorSpace.supportsOutput() ? sourceColorSpace : DestinationColorSpace::SRGB();
     for (auto length : lengths) {
-        RetainPtr context = adoptCF(CGBitmapContextCreate(nullptr, length, length, 8, length * 4, colorSpace.get(), kCGImageAlphaPremultipliedLast));
-        if (!context)
-            return nullptr;
-
-        // Quality is more important than rendering speed for the current use case.
-        CGContextSetInterpolationQuality(context.get(), kCGInterpolationHigh);
-        CGContextDrawImage(context.get(), CGRectMake(0, 0, length, length), image.platformImage().get());
-        RetainPtr newImage = adoptCF(CGBitmapContextCreateImage(context.get()));
-        if (!newImage)
-            return nullptr;
-
-        CGImageDestinationAddImage(destination.get(), newImage.get(), nullptr);
+        if (RefPtr bitmap = ShareableBitmap::createFromImageDraw(image, destinationColorSpace, { (int)length, (int)length }, ShareableBitmap::ShouldShrinkToFit::Yes)) {
+            bitmaps.append(WTFMove(bitmap));
+            continue;
+        }
+        return { };
     }
 
-    if (!CGImageDestinationFinalize(destination.get()))
-        return nullptr;
-
-    return SharedBuffer::create(destinationData.get());
+    return bitmaps;
 }
 
-static RefPtr<SharedBuffer> expandSVGImageToData(SVGImage& svgImage, ASCIILiteral uti, std::span<const unsigned> lengths)
+static RefPtr<NativeImage> nativeImageFromSVGImage(SVGImage& image, unsigned length)
```

Comment:
I wonder if we want to be clearer here that this always returns a square/icon. But since it's a static and pretty obvious it doesn't matter too much I think.